core/src/main/java/org/elasticsearch/node/ResponseCollectorService.java

```diff
@@ -0,0 +1,144 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.node;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.ExponentiallyWeightedMovingAverage;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+/**
+ * Collects statistics about queue size, response time, and service time of
+ * tasks executed on each node, making the EWMA of the values available to the
+ * coordinating node.
+ */
+public class ResponseCollectorService extends AbstractComponent implements ClusterStateListener {
+
+    private static final double ALPHA = 0.3;
+
+    private final ConcurrentMap<String, NodeStatistics> nodeIdToStats = ConcurrentCollections.newConcurrentMap();
+
+    public ResponseCollectorService(Settings settings, ClusterService clusterService) {
+        super(settings);
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.nodesRemoved()) {
+            for (DiscoveryNode removedNode : event.nodesDelta().removedNodes()) {
+                removeNode(removedNode.getId());
+            }
+        }
+    }
+
+    void removeNode(String nodeId) {
+        nodeIdToStats.remove(nodeId);
+    }
+
+    public void addNodeStatistics(String nodeId, int queueSize, long responseTimeNanos, long avgServiceTimeNanos) {
+        NodeStatistics nodeStats = nodeIdToStats.get(nodeId);
+        nodeIdToStats.compute(nodeId, (id, ns) -> {
+            if (ns == null) {
+                ExponentiallyWeightedMovingAverage queueEWMA = new ExponentiallyWeightedMovingAverage(ALPHA, queueSize);
+                ExponentiallyWeightedMovingAverage responseEWMA = new ExponentiallyWeightedMovingAverage(ALPHA, responseTimeNanos);
+                NodeStatistics newStats = new NodeStatistics(nodeId, queueEWMA, responseEWMA, avgServiceTimeNanos);
+                return newStats;
+            } else {
+                ns.queueSize.addValue((double) queueSize);
+                ns.responseTime.addValue((double) responseTimeNanos);
+                ns.serviceTime = avgServiceTimeNanos;
+                return ns;
+            }
+        });
+    }
+
+    public Map<String, ComputedNodeStats> getAllNodeStatistics() {
```

Comment:
I'm planning on using this method for the very next PR that implements the actual ranking algorithm, without this method the `NodeStatistics` key would be exposed, which is the mutable one. By putting this method here `NodeStatistics` can be private and we can expose only the immutable `ComputedNodeStats`